### PR TITLE
Move `trainer.{fit,validate,test,predict,tune}` to the top of `trainer.py`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Move `trainer.{fit,validate,test,predict,tune}` to the top of `trainer.py`
**NO OTHER CHANGES**
Fixes https://github.com/PyTorchLightning/pytorch-lightning/pull/7232#discussion_r621800173

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [n/a] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by kaushikb11 — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
Love this!

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -410,6 +410,279 @@ def __init__(
...
+        """
+        Trainer._log_api_event("fit")
+
+        self.state = TrainerState.FITTING
+        self.training = True
+
+        # if a datamodule comes in as the second arg, then fix it for the user
+        if isinstance(train_dataloader, LightningDataModule):
+            datamodule = train_dataloader
+            train_dataloader = None
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:443]: yikes - i know its legacy behavior but this is why we have typehints 😞

**Review by s-rog — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:443]: What are you trying to say @ananthsub?. Do you prefer that overloading the `train_dataloaderr` argument by passing a `DataModule` positionally is not allowed?
